src/Native/Runtime/arm/WriteBarriers.S

```diff
@@ -2,67 +2,375 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-// TODO: Implement Unix write barriers
-#include <unixasmmacros.inc>
-
 .syntax unified
 .thumb
 
-LEAF_ENTRY RhpAssignRef, _TEXT
+#include <unixasmmacros.inc>
+#include <AsmOffsets.inc>         // generated by the build from AsmOffsets.cpp
+
+#ifdef WRITE_BARRIER_CHECK
+
+.macro UPDATE_GC_SHADOW BASENAME, REFREG, DESTREG
+
+          // If g_GCShadow is 0, don't perform the check.
+          ldr          r12, =C_FUNC(g_GCShadow)
+          ldr          r12, [r12]
+          cbz          r12, LOCAL_LABEL(\BASENAME\()_UpdateShadowHeap_Done_\REFREG)
+
+          // Save DESTREG since we're about to modify it (and we need the original value both within the macro and
+          // once we exit the macro). Note that this is naughty since we're altering the stack pointer outside of
+          // the prolog inside a method without a frame. But given that this is only debug code and generally we
+          // shouldn't be walking the stack at this point it seems preferable to recoding the all the barrier
+          // variants to set up frames. The compiler knows exactly which registers are trashed in the simple write
+          // barrier case, so we don't have any more scratch registers to play with (and doing so would only make
+          // things harder if at a later stage we want to allow multiple barrier versions based on the input
+          // registers).
+          push         \DESTREG
+
+          // Transform DESTREG into the equivalent address in the shadow heap.
+          ldr          r12, =C_FUNC(g_lowest_address)
+          ldr          r12, [r12]
+          sub          \DESTREG, r12
+          cmp          \DESTREG, #0
+          blo          LOCAL_LABEL(\BASENAME\()_UpdateShadowHeap_PopThenDone_\REFREG)
+          ldr          r12, =C_FUNC(g_GCShadow)
+          ldr          r12, [r12]
+          add          \DESTREG, r12
+          ldr          r12, =C_FUNC(g_GCShadowEnd)
+          ldr          r12, [r12]
+          cmp          \DESTREG, r12
+          jhi          LOCAL_LABEL(\BASENAME\()_UpdateShadowHeap_PopThenDone_\REFREG)
+
+          // Update the shadow heap.
+          str          \REFREG, [\DESTREG]
+
+          // The following read must be strongly ordered wrt to the write we've just performed in order to
+          // prevent race conditions.
+          dmb
+
+          // Now check that the real heap location still contains the value we just wrote into the shadow heap. This
+          // read must be strongly ordered wrt to the previous write to prevent race conditions. We also need to
+          // recover the old value of DESTREG for the comparison so use an xchg instruction (which has an implicit lock
```

Comment:
`xchg` is x64-specific instruction. It does not seem to be used here. Should this rather match https://github.com/dotnet/corert/blob/master/src/Native/Runtime/arm/WriteBarriers.asm#L82 ?